File: problems/Python/29. Divide Two Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        imax = 2**31 - 1
        imin = -2**31

        if dividend == imin and divisor == -1:
            return imax

        sgn = -1 if (dividend < 0) ^ (divisor < 0) else 1
        dvd = abs(dividend)
        dvs = abs(divisor)
        res = 0
        
        while dvd >= dvs:
            temp, mult = dvs, 1
            while dvd >= (temp << 1):
                temp <<= 1
                mult <<= 1
            dvd -= temp
            res += mult
            
        return sgn * res

File: problems/Java/29. Divide Two Integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }

        int sgn = (dividend < 0) ^ (divisor < 0) ? -1 : 1;
        long dvd = Math.abs((long) dividend);
        long dvs = Math.abs((long) divisor);
        long res = 0;

        while (dvd >= dvs) {
            long temp = dvs, mult = 1;
            while (dvd >= (temp << 1)) {
                temp <<= 1;
                mult <<= 1;
            }
            dvd -= temp;
            res += mult;
        }

        return (int) (sgn * res);
    }
}
